Maneuver fm

mly32 opened the pull request:
### Summary
<!-- What changes were made? What features were added? What bugs were fixed? -->
This PR addresses Jira ticket CISLUNAR-18.

Maneuver mode was updated to check for glow plug failure.

### Testing
<!-- How was your code tested? Include locations of test code and/or documents -->

Lacks testing at this stage.

### Notes
<!--- List any major or minor points, future thoughts, and/or future concerns -->

Is it possible to have the maneuver mode's initialization defined outside of `main`? It feels like it really belongs in maneuver_flightmode. Say I end up changing the queue type (e.g., I use a heap instead of python's default priority_queue class, or like in this PR where I changed it from a queue to a priority_queue). Ideally, none of the other functions/parts of the code would need to know, they would just use ManeuverMode().addEvent() or something similar. 

The queue and other data still are not saved onto disk in the event of a reboot. 

### Comments
- [x] I'm not sure why set_parameter uses kwargs and how to use the function defined in command_defintions. 
- [x] What is the behavior for when a scheduled burn time has already passed without maneuver mode running?
- [x] I left the constants as basic values, but they need to be updated.
- [x] My glowplug check logic and glowplug failure logics are basic and could use improvements.

tmf97 reviewed (dismissed):
Looks really really good - We talked about changing the maneuver queue to a list or something else - up to you on how to implement this.

mly32 commented:
Added (basic) tests using mocks. Fixed/updated maneuver mode and moved params into paramaters.py.

Failing the CI right now because I placed the new params in the wrong place.

tmf97 commented on flight_modes/maneuver_flightmode.py:
```diff
@@ -0,0 +1,78 @@
+            prior_pressure = self.get_pressure()
+            print(prior_pressure)
+            if self.glowplug1_valid:
+                self._parent.gom.glowplug(consts.GLOWPLUG_DURATION)
```
Oh, we should be using `parameters.GLOWPLUG_DURATION` instead of `constants...`, my bad for not seeing that earlier. 
The following should all be removed from constants.py, and used from parameters instead:
```
SPLIT_BURNWIRE_DURATION = 1  # second
ANTENNAE_BURNWIRE_DURATION = 1  # second
GLOWPLUG_DURATION = 1  # SECOND
BURN_WAIT_TIME = 15  # minutes
```

The reason for this is that it's very easy to change the values of parameters (there's a command where we can uplink a new value), but very hard to change the value of a parameter (we have to uplink a new `constants.py` file containing the new value, then restart the flight software) - it's doable, just much more involved.

tmf97 commented on main.py:
```diff
@@ -54,7 +55,7 @@ def __init__(self):
         self.reorientation_queue = Queue()
         self.reorientation_list = []
-        self.maneuver_queue = Queue()  # maneuver queue
+        self.maneuver_queue = PriorityQueue()
```
Weren't we talking about changing the maneuver_queue to a list?

tmf97 replied:
Huh, I learned a thing or two from https://stackoverflow.com/questions/9287919/can-i-get-an-item-from-a-priorityqueue-without-removing-it-yet

tmf97 commented on flight_modes/maneuver_flightmode.py:
```diff
@@ -0,0 +1,78 @@
+                self.task_completed = self.valid_glowplug(current_pressure, prior_pressure)
+                self.glowplug2_valid = self.task_completed
+            if not self.glowplug1_valid and not self.glowplug2_valid:
+                # TODO ground msg?... do a final check?
```
I've thought about defining a set of "error codes" that we could collect and downlink on request. In this case we'd add something to the error code "queue" or something. Defining that is for another PR though, just keep the TODO for now.
Something we could do here is to try to increase the GLOWPLUG_DURATION, but I think that increase  would be best performed by ground operators.

tmf97 commented on flight_modes/maneuver_flightmode.py:
```diff
@@ -0,0 +1,78 @@
+        # prepare next maneuver
+        smallest_time_burn = -1
+        while not self._parent.maneuver_queue.empty():
+            smallest_time_burn = self._parent.maneuver_queue.get()
```
Remember that `queue.get()` removes the item from the queue as well as returning it - are you sure you want that here?

tmf97 commented on flight_modes/maneuver_flightmode.py:
```diff
@@ -0,0 +1,78 @@
+            else:
+                break
+
+        set_parameter("SCHEDULED_BURN_TIME", smallest_time_burn, False)  # for testing
```
Same comment as in command_definitions.py

tmf97 commented on utils/constants.py:
```diff
@@ -194,9 +194,14 @@
+GLOWPLUG_DURATION = 1  # SECOND
+GLOW_WAIT_TIME = 2
+PRESSURE_THRESHOLD = 1000
+PRESSURE_DELTA = -1
```
All these should be in parameters, not constants

tmf97 commented on tests/maneuver_test.py:
```diff
@@ -0,0 +1,47 @@
+consts.GLOW_WAIT_TIME = 0.01
+
+
+class FakeGOM():
```
This is something we'll eventually want to breakout into its own file, but good enough for now

tmf97 reviewed (changes requested):
Almost there - all the logic/reasoning is perfect, just some implementation quirks need ironing out. Huge fan of the maneuver test!

mly32 commented:
constants need to be updated since there are still other variables that should be parameters such as `SPLIT_BURNWIRE_DURATION`.

I don't like using a string to set parameters as done here: `set_parameter("GLOWPLUG2_VALID", self.task_completed, consts.FOR_FLIGHT)`. I will look into a way to do this safely in another PR.

Will also look into removing parameters.json as part of the codebase. it should be something that is git-ignored and should be generated if the file does not exist

tmf97 commented on tests/maneuver_test.py:
```diff
@@ -0,0 +1,51 @@
+params.GLOW_WAIT_TIME = 0.01
+
+
+class FakeGOM:
+    def glowplug(self, x):
+        return 0
+
+    def glowplug2(self, x):
+        return 0
```
Really jank, but this worked on my machine:
```suggestion
class FakeGOM(Gomspace):

    def __init__(self):
        self.pc = None

    def glowplug(self, duration, delay=0):
        return None

    def glowplug2(self, duration, delay=0):
        return None
```

tmf97 reviewed (approved):
Looks solid! We should flatsat test this soon - how does this weekend work?

mly32 commented:
> Looks solid! We should flatsat test this soon - how does this weekend work?

Sounds like a plan!
